Implement to support shot-vector to Catalyst

sengthai opened the pull request:
**Context:**
Implement to support shot-vector to Catalyst in one or multiple sample measurements and add to ensure its functionality.

**Description of the Change:**
- Implement to support for a shot-vector in Jax tracer
   - Add helpers two functions for getting the total number of copies and shot vector
   - Modify the trace measurement process to support shot-vector 
- Add to ensure the its functionality.

**Related GitHub Issues:**
PennyLaneAI#980

rmoyard commented:
Hello @sengthai, thanks for the PR! I will take a look tomorrow.

sengthai commented:
> Hello @sengthai, thanks for the PR! I will take a look tomorrow.

Thank you @rmoyard, I appreciate it! Let me know if you need any additional information. 

Regarding #980 , the element form of the shot vector `((20, 5))` should be equivalent to `(20,) * 5`, which expands to `(20, 20, 20, 20, 20)`. However, in the current implementation, `((20, 5),)` is treated as equivalent to `(20,) * 5`.

If we want to align with the expected behavior where `((20, 5))` is interpreted as `(20,) * 5`, we can update the [Shots](https://docs.pennylane.ai/en/stable/_modules/pennylane/measurements/shots.html#Shots) class in PennyLane accordingly.

dime10 commented on doc/changelog.md:
````diff
@@ -18,6 +18,30 @@ This release contains contributions from (in alphabetical order):
+  Array([[0], [1], [1], [0], [1]], dtype=int64),
+  Array([[1], [0], [1], [1], [0], [1],[0]], dtype=int64))
+  ```
+
````
You can also include your name in the contributor list at the end of the current changelog section :)

dime10 commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -859,6 +874,15 @@ def trace_quantum_measurements(
     for i, o in enumerate(outputs):
         if isinstance(o, MeasurementProcess):
+
+            # Check if the measurement is supported shot-vector where num_of_total_copies > 1
+            num_of_total_copies = get_device_total_shot_copies(device)
+            if num_of_total_copies > 1 and o.return_type.value != "sample":  # qml.sample()
+                raise NotImplementedError(
```
Nice!

Looks like this check is missing test coverage however. Typically we verify errors as well using pytest.

sengthai replied:
Good catch! I'll add test coverage for this check. Thanks for pointing that out!

dime10 commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -869,19 +893,31 @@ def trace_quantum_measurements(
+                    else:
+                        shape = (shots, nqubits) if using_compbasis else (shots,)
+
+                        for _ in range(copies):
+                            result = sample_p.bind(obs_tracers, shots=shots, shape=shape)
```
Nice, I think this will work!

However, on some devices, having multiple sample instructions like this could lead to multiple executions of the circuit. Can you think of a way to handle the shot vectors with only a single sample call?

sengthai replied:
Thanks! That's a great point. @rmoyard offered a helpful hint of applying some post processing. I think we can create a single call that requests the total number of shots needed. For example, if we need 3, 3, 3, 3, 2, 4, 4 shots (shot vector `((3,4), 2, (4,2))`), we would aggregate these into a single call for 22 shots.  The execution result will be a tensor containing all the samples. Then we apply some post-processing by splitting the tensor into several sub-tensors according to the required shot copies. Does this approach sound feasible?

dime10 replied:
Nice, that's what I was thinking of :)

dime10 reviewed (commented):
Thank you for submitting your solution @sengthai :)

> Regarding https://github.com/PennyLaneAI/catalyst/issues/980 , the element form of the shot vector ((20, 5)) should be equivalent to (20,) * 5, which expands to (20, 20, 20, 20, 20). However, in the current implementation, ((20, 5),) is treated as equivalent to (20,) * 5.

Oh my bad, I think this is just a typo. It should have been `((20, 5),)`.

rmoyard commented on doc/changelog.md:
```diff
@@ -18,6 +18,30 @@ This release contains contributions from (in alphabetical order):
 
 <h3>New features</h3>
 
+* Shot-vector support for Catalyst: Introduces support for shot-vectors in Catalyst, currently available for `qml.sample` measurements in the `lightning.qubit` device. Shot-vectors now allow elements of the form `((20, 5),1)`, which is equivalent to `(20,)*5` or `(20, 20, 20, 20, 20)`. Furthermore, multiple `qml.sample` calls can now be returned from the same program, and can be structured using  ...
```
Yes I agree with David, it is 5 times 20 shots followed by one shot

sengthai replied:
Thanks for the input! That's my mistake. It was a typo. It should be `((20, 5),)` in this changelog.

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -147,6 +147,21 @@ def get_device_shots(dev):
+    """Helper function to get device shot vector."""
+
+    if isinstance(dev, qml.devices.LegacyDevice):
+        return [(dev.shots, 1)]
```
I am not completely sure it is correct  for the old device API. I will double check

sengthai replied:
Got it, thanks for double-checking! Let me know what you find, and I will adjust it as needed.

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -147,6 +147,21 @@ def get_device_shots(dev):
+    if isinstance(dev, qml.devices.LegacyDevice):
+        return [(dev.shots, 1)]
+
+    shot_vector = [(shot_copy.shots, shot_copy.copies) for shot_copy in dev.shots.shot_vector]
```
:+1:

rmoyard commented on frontend/test/pytest/test_shotvector.py:
```diff
@@ -0,0 +1,69 @@
+from catalyst import qjit
+
+
+class TestShotVector:
```
Could you add a test for a shotvector like ((20,5), 100, (101,2)) that mixes shots with and without copies?

sengthai replied:
Sure, I’ll add a test for it. Thanks for the suggestion!

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -1225,9 +1261,24 @@ def is_leaf(obj):
                 meas_results = tree_unflatten(meas_trees, meas_tracers)
 
+                # Always return a tuple of results with shot-vectors where num_of_total_copies > 1
+                if get_device_total_shot_copies(device) > 1:
```
Minor: here I think you can handle any shape of return by using pytrees of the return values. https://jax.readthedocs.io/en/latest/pytrees.html You can then handle any return shape even something that mixes dict and list. "{"first": qml.sample(), "second": [100, qml.sample()]}".  They are multiple place in the codebase where we use that.

sengthai replied:
Thanks for the tip! It sounds like a robust solution. I’ll look into integrating this approach. Appreciate the guidance!

dime10 replied:
Just a note that arbitrary output containers are already being restored on the line just above:
`meas_results = tree_unflatten(meas_trees, meas_tracers)`. 

The processing you are doing below looks to me like it just replaces lists with tuples (bare and inside one layer of dictionaries). I think this is unnecessary, and is probably better handled by just replacing the list you produce in `trace_quantum_measurements` with a tuple directly in that function.

You can still add the test case Romain suggested to make sure more complex container structures will work though :)

rmoyard reviewed (commented):
Great job @sengthai ! As David mentioned, you can probably handle the shot vector in one execution and then apply some post processing. On a more minor issue I think you can also handle more return types (nested one e.g.) by using pytress.

sengthai commented:
Hello, @dime10 and @rmoyard. I've updated this to support a shot vector with a single sample call. And I’ve added the necessary tests and cleaned up some redundant code from the previous implementation. I appreciate your time and feedback—thank you both!

dime10 commented:
> Hello, @dime10 and @rmoyard. I've updated this to support a shot vector with a single sample call. And I’ve added the necessary tests and cleaned up some redundant code from the previous implementation. I appreciate your time and feedback—thank you both!

Nice work 🎉

rmoyard commented on doc/changelog.md:
```diff
@@ -0,0 +1,3992 @@
+
+This release contains contributions from (in alphabetical order):
+
+Sengthai Heng,
```
We have moved the changelog to `doc/dev/releases` in changelog-dev, could you move it and erase this file?

rmoyard replied:
Yes correct !

rmoyard commented on frontend/test/pytest/test_shotvector.py:
```diff
@@ -0,0 +1,162 @@
+        assert jnp.shape(circuit_dict()["first"]) == (4, 3, 1)
+        assert jnp.shape(circuit_dict()["second"]) == (4, 3, 1)
+
+    def test_shot_vector_with_mixes_shots_and_without_copies(self):
```
Great :+1:

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -146,6 +146,16 @@ def get_device_shots(dev):
 
+def get_device_total_shot_copies(dev):
+    """Helper function to get device total shot copies."""
+    return 1 if isinstance(dev, qml.devices.LegacyDevice) else dev.shots.num_copies
```
```suggestion
    return dev.shots.num_copies
```
PennyLane has created a facade that should remove the calling different method for the devices.

dime10 replied:
I guess in that case the helper could be removed entirely :)

sengthai replied:
Good point! Since PennyLane's facade now standardizes the method calls for devices, removing the helper function makes sense. I'll go ahead and refactor that part. Thanks for the suggestion!

rmoyard commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -870,7 +891,21 @@ def trace_quantum_measurements(
+                    if len(reshaped_result) == 1:
+                        reshaped_result = reshaped_result[0]
+
+                    out_classical_tracers.append(reshaped_result)
```
The reshape looks good, I let @dime10 double check :+1:

dime10 replied:
Looks great to me 🎉

rmoyard reviewed (commented):
Great job, happy to approve after some minor changes!

rmoyard reviewed (approved):
Thanks for your contribution, great job :100:

sengthai commented:
> Thanks for your contribution, great job 💯

Thank you both for the kind words! I really appreciate your feedback and support. Excited to continue contributing!

rmoyard commented:
Hi @sengthai I have merged your PR

sengthai commented:
> Hi @sengthai I have merged your PR

Thank you for merging the PR!
